src/puppetlabs/puppetdb/catalogs.clj

```diff
@@ -158,7 +158,9 @@
    :file (s/maybe s/Str)
    :line (s/maybe s/Int)
    :parameters {s/Any s/Any}
-   :resource s/Str
+   ;; On historical-catalogs this is an optional-key but this is required on
+   ;; "normal" catalogs
+   (s/optional-key :resource) s/Str
```

Comment:
Either way I'd like to address it in another patch, it requires some nontrivial changes to the storage code.